milli/src/update/settings.rs

```diff
@@ -465,14 +473,23 @@ impl<'a, 't, 'u, 'i> Settings<'a, 't, 'u, 'i> {
     fn update_exact_attributes(&mut self) -> Result<bool> {
         match self.exact_attributes {
             Setting::Set(ref attrs) => {
-                let attrs = attrs.iter().map(String::as_str).collect::<Vec<_>>();
-                self.index.put_exact_attributes(self.wtxn, &attrs)?;
-                Ok(true)
-            }
-            Setting::Reset => {
-                self.index.delete_exact_attributes(self.wtxn)?;
-                Ok(true)
+                let old_attrs = self
+                    .index
+                    .exact_attributes(self.wtxn)?
+                    .iter()
+                    .cloned()
+                    .map(String::from)
+                    .collect::<HashSet<String>>();
+
+                if attrs != &old_attrs {
+                    let attrs = attrs.iter().map(String::as_str).collect::<Vec<_>>();
+                    self.index.put_exact_attributes(self.wtxn, &attrs)?;
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
```

Comment:
And you are doing exactly the same here, right?

Reply:
Correct. Here I compare the two `HashSet`s of old/new, whereas up above I compared the two `Vec`s of old/new.

Also, if you would mind taking a look at [this comment above](https://github.com/meilisearch/milli/pull/706#issuecomment-1325885217), I would really appreciate it. Not sure if that assumption I made for the `Settings::Reset` case is correct.

Reply:
Thanks for the answer (and correcting that example I gave--I overlooked the `&&` behavior); however, I think I wasn't specific enough:

Currently, `delete_all_searchable_fields`:
- returns the result of executing `self.delete_user_defined_searchable_fields(wtxn)`
- ignores the result of executing `self.delete_searchable_fields(wtxn)?;`

When should reindexing happen (logically speaking here)? Here are a few options that I thought of:
- Whenever `delete_searchable_fields` returns `true`?
- when `delete_user_defined_searchable_fields` and `delete_searchable_fields` both return true?
- when at least one of the above return true?

I am not versed well-enough in milli's inner-workings to be able to answer this question myself, as I am not 100% clear how `user_defined_searchable_fields` and `searchable_fields` relate to reindexing.

Reply:
I would say _when at least one of the above returns `true`_. Thank you very much for your work on the engine BTW, very pleasant to see you involved :)

Reply:
This is in relation to optimizing the `Settings::Reset` behavior--if a reset was selected, but no reset is needed, then we shouldn't need to reindex. However, we also have to be careful to not miss a reindex with `Settings::Reset` when we actually need one. We could be safe and return an `||` of the results of the two above methods, but this could cause reindexes when they aren't needed.

Reply:
> I would say _when at least one of the above returns `true`_. Thank you very much for your work on the engine BTW, very pleasant to see you involved :)

@Kerollmops Thanks! I'll apply the fix for that in a couple hours.

Reply:
I made the appropriate change in 87e2bc3.

@irevoire I am roping you in here as an extra pair of eyes because of your work in f586028f9a8250e462c43484c6ce4867dc2d5aff 6 months ago. Does my new implementation of `delete_all_searchable_fields` in 87e2bc3 return true *only* when reindexing needs to occur? I am trying hard not to mess this up because I don't want to cause reindexing when it is actually not needed.

If I have it correct in 87e2bc3, maybe we could also add a check in to see if the user had defined any fields previously and only call `delete_searchable_fields`/`delete_all_searchable_fields` when the user had previously defined searchable fields?

Reply:
Hey @GregoryConrad,
> Does my new implementation of delete_all_searchable_fields in https://github.com/meilisearch/milli/commit/87e2bc3beda573a72c87647d614393d524bd5b44 return true only when reindexing needs to occur?

Yep, I think it should be ok like that.

> If I have it correct in https://github.com/meilisearch/milli/commit/87e2bc3beda573a72c87647d614393d524bd5b44, maybe we could also add a check in to see if the user had defined any fields previously and only call `delete_searchable_fields`/`delete_all_searchable_fields` when the user had previously defined searchable fields?

I'm not sure I see the point in doing that. If the user had no searchable_field defined, then calling `delete_searchable_fields` or `delete_all_searchable_fields` will return `false` and won't trigger a re-indexing, right?

Reply:
> If the user had no searchable_field defined, then calling delete_searchable_fields or delete_all_searchable_fields will return false and won't trigger a re-indexing, right?

I wasn't sure if `delete_searchable_fields` *actually* would return false in that case, which is why I brought it up. Reason: it seemed like `searchable_fields` references the currently searchable fields, which will always exist, and wiping those out would need a reindex (even if you just populate the same searchable fields again in the index). Thus, the idea for the optimization. I can experiment with the current behavior in a little bit and see for sure how the index acts

Reply:
Looks like we're all set as-is:
```rust
        println!("START");
        let index = TempIndex::new();

        index
            .add_documents(documents!([
                { "id": 1, "name": "kevin", "age": 23 },
                { "id": 2, "name": "kevina", "age": 21},
                { "id": 3, "name": "benoit", "age": 34 }
            ]))
            .unwrap();

        index.update_settings(|settings| {
            settings.set_searchable_fields(vec!("name".to_string()));
        }).unwrap();
        index.update_settings(|settings| {
            settings.reset_searchable_fields();
        }).unwrap();
        index.update_settings(|settings| {
            settings.reset_searchable_fields();
        }).unwrap();

        println!("FINISH");
```
Yields the following output:
```
START
REINDEXING!
REINDEXING!
FINISH
```
Which is expected (2 reindexes). I was unsure if the second `reset_searchable_fields` call would cause an extra reindex in the implementation as-is.

Reply:
Yep, I think this process is now totally _clear_ to you! :grin: 

Since you went pretty far in understanding how it works, it doesn't impact your PR, but maybe you would be interested in knowing how the « real » searchable fields work compared to the user-defined ones.

Basically, IIRC, if you set `doggo` as a searchable field and have this document:
```JSON
{
  "id": 1,
  "doggo": {
    "name": "bob",
    "race": "Bernese mountain dog"
  }
}
```

Then your user-defined searchable fields are going to be `doggo`, while the real searchable fields are going to be `doggo.name, doggo.race` (and maybe even `doggo`, I don't remember).
And this second list might update itself while you add more documents.
If you add a second document like that:
```JSON
{
  "id": 2,
  "doggo": {
    "name": "jean",
    "race": "golden retriever",
    "comment": "Jean is an adorable golden retriever that loves to be cuddled."
  }
}
```

Then while your user-defined searchable fields still won't change, your real/hidden searchable fields will add a third `doggo.comment` field.

Reply:
Thanks for that amazing explanation @irevoire !